### PR TITLE
Refactor EmberCLI to run Build Pipeline, Express Server & Testem Server in separate processes & corresponding addon API

### PR DESCRIPTION
**Pull request by taras**
[RFC](https://github.com/embersherpa/rfcs/blob/master/active/0000-multiprocess-ember-cli.md)

### CODE REVIEW

**Comment by taras** [comment]
@bcardarella @stefanpenner @abuiles @rwjblue What do you guys think?

File: active/0000-multiprocess-ember-cli.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+		]
+	}
+}
+```
+
+The processes will be automatically stoped when ember-cli is stopped.
+
+# Drawbacks
+
+I don't know.
```
Comment (stefanpenner) [inline active/0000-multiprocess-ember-cli.md]: - complexity
- cross platform process management (don't want orphans)
  *

File: active/0000-multiprocess-ember-cli.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+# Alternatives
+
+Please, suggest.
```
Comment (stefanpenner) [inline active/0000-multiprocess-ember-cli.md]: stricter internal concept for isolation, or exploration stuff in node 0.11.x and iojs .. I believe @domenic worked on something related, but i maybe be mistaken.

File: active/0000-multiprocess-ember-cli.md (hunk 1 above)
Comment (taras) [inline active/0000-multiprocess-ember-cli.md]: Ok, I will add these 2 points.

I think `node-foreman` might handle the multi platform consideration for us.

Comment (stefanpenner) [inline active/0000-multiprocess-ember-cli.md]: we should investigate how well it works.

**Comment by stefanpenner** [comment]
this also replaces the current hacky reloading stuff we do.

File: active/0000-multiprocess-ember-cli.md (hunk 1 above)
Comment (taras) [inline active/0000-multiprocess-ember-cli.md]: I can make a prototype and we can try it on windows.

**Comment by stefanpenner** [comment]
I would love to visit this again post 1.0, It will need some investigation but still am +1

**Comment by stefanpenner** [comment]
I like where this is going, but this needs more fleshing out, specifically.
1. error handling (processes fail, either parent or child etc), we need to be sure they crash correctly, and cleanup occurs correctly.
2. what does the process tree look like?
3. ability to reload reprocesses etc.

**Comment by nathanhammond** [comment]
@taras et al.:

The ability to spawn and manage processes as well as pipe information from one process into the next is something that we want inside of ember-cli. This is a required component to be able to build Ember applications and have them play nicely with things like NW.js, Electron, Cordova, and more, many of which have their own specific tools which we need to integrate with gracefully.

For the creating processes themselves, we would like for them to be achievable as an ember-cli addon, as well as being addressable from other "peerDependency" addons. We'd then delegate back to the `ember` invocation at specific points in the lifecycle in order to allow ember-cli to manage the processes.

So, with that, we do want this feature, but we'd like this RFC updated to address specific implementation details for arbitrary process counts, process relationships, bidirectional communication, references to the process itself, failure recovery, and things we've not yet discovered since we haven't tried to complete this design work.

This is part of our "core + extensibility" efforts.

**Comment by stefanpenner** [comment]
> I like where this is going, but this needs more fleshing out, specifically.
> 
> error handling (processes fail, either parent or child etc), we need to be sure they crash correctly, and cleanup occurs correctly.
> what does the process tree look like?
> ability to reload reprocesses etc.

^^ should be addressed for this RFC to help us explore the solution here.

**Comment by Turbo87** [comment]
closing due to inactivity 😢

**Pull request closed without merging**